---
title: "NCR Primary Guide Screening"
author: "Amanda Mok"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: html_document
---

```{r setup, echo=F, include=F}
library(here)
library(reshape)
library(ggplot2)
library(patchwork)
library(nlme)
library(Biostrings)
library(bedr)
library(factoextra)
library(ggseqlogo)

knitr::opts_chunk$set(echo=F, cache=T, fig.width=8)

compute_rate <- function(data_subset, 
                         x="time",
                         y="signal_bkgdSubtract") {
  # screening_data: data.frame
  # x: character; column name to regress against
  # y: character; column name to regress
  model_formula <- formula(paste(y, "~", x))
  model_fit <- lm(model_formula, data_subset)
  model_rate <- coef(model_fit)[x]
  return(model_rate)
}

analyze_guide <- function(screening_data, guide,
                          signal="signal_bkgdSubtract",
                          time_start=15, mixed_model=F) {
  # screening_data: data.frame
  # guide: character
  # model: formula to be used for lm() regression
  # time_start: timepoint to start regression
  # mixed_model: logical; whether to run linear regression with random effects for 384-well
  data_subset <- subset(screening_data, guide_id==guide)
  data_subset$activator <- relevel(data_subset$activator, ref="noActivator")
  time_stop <- signif(max(data_subset$time))
  if(!mixed_model) {
    # compute fit
    model_fit <- lm(formula(paste(signal, "~ time*activator")), data=data_subset)
    # predict trajectory
    model_fit_coef <- coef(model_fit)
    data_predict <- expand.grid(time=seq(time_start, time_stop, 1),
                                activator=c("noActivator", "100fM"))
    data_predict[[signal]] <- predict(model_fit, newdata=data_predict)
    # pull coefficients
    guide_coef <- data.frame(summary(model_fit)$coefficients, guide_id=guide,
                             coef=rownames(summary(model_fit)$coefficients), row.names=NULL)
    colnames(guide_coef)[colnames(guide_coef)=="Pr...t.."] <- "p.value"
    colnames(guide_coef)[colnames(guide_coef)=="Std..Error"] <- "Std.Error"
  } else {
    # compute fit
    model_fit <- nlme::lme(formula(paste(signal, "~ time*activator")),
                           random = ~ 1 + time | well_384, 
                           data = data_subset)
    # predict trajectory
    model_fit_coef <- summary(model_fit)$tTable[,1]
    data_predict_noActivator <- data.frame(time=seq(time_start, time_stop, 1),
                                           activator="noActivator")
    data_predict_noActivator[[signal]] <- model_fit_coef["(Intercept)"] + 
      data_predict_noActivator$time * model_fit_coef["time"]
    data_predict_100fM <- data.frame(time=seq(time_start, time_stop, 1),
                                     activator="100fM")
    data_predict_100fM[[signal]] <- model_fit_coef["(Intercept)"] + 
      model_fit_coef["activator100fM"] +
      data_predict_100fM$time * (model_fit_coef["time"] + model_fit_coef["time:activator100fM"])
    data_predict <- rbind(data_predict_noActivator, data_predict_100fM)
    # pull fixed effects
    guide_coef <- data.frame(summary(model_fit)$tTable, guide_id=guide,
                             coef=rownames(summary(model_fit)$varFix), row.names=NULL)
    colnames(guide_coef)[colnames(guide_coef)=="Value"] <- "Estimate"
    guide_coef <- guide_coef[, -which(colnames(guide_coef)=="DF")]
  }
  # make plot
  plot_max <- max(data_subset[[signal]], na.rm=T)
  plot_width <- max(data_subset[[signal]], na.rm=T) - min(data_subset[[signal]], na.rm=T)
  rate_noActivator <- model_fit_coef["time"]
  rate_100fM <- model_fit_coef["time"] + model_fit_coef["time:activator100fM"]
  plot_text <- data.frame(x=10,
                          y=plot_max - c(1:3)*plot_width/20,
                          text=c(paste("no activator rate:", signif(rate_noActivator, 3)),
                                 paste("100 fM rate:", signif(rate_100fM, 3)),
                                 paste("p =", 
                                       signif(subset(guide_coef, coef=="time:activator100fM")$p.value, 3))))
  guide_plot <- ggplot(data_predict, 
                       aes_string(x="time", y=signal, col="activator")) +
    geom_line(size=3) + 
    geom_point(data=data_subset, 
               aes_string(x="time", y=signal, col="activator"), alpha=0.5) +
    geom_text(data=plot_text, aes(x=x, y=y, label=text), col="black", hjust=0) +
    theme_bw() + scale_color_manual(values=c("red", "black")) +
    ggtitle(paste0("NCR_", guide)) + xlab("time (min)") + ylab("signal (RFU)")
  return(list(plot=guide_plot, coef=guide_coef))
}

load_sam <- function(sam_fname) {
  # sam_fname: character; file.path to sam alignment file
  sam_colnames <- c("qname", "flag", "rname", "pos", "mapq",
                  "cigar", "rnext", "pnext", "tlen", "seq", "qual")
  sam_data <- read.table(sam_fname, comment.char="@")
  colnames(sam_data)[seq(length(sam_colnames))] <- sam_colnames
  return(sam_data)
}

plot_guide_feature_cor <- function(plot_data, x, x_lab, y, y_lab, in_vivo=F) {
  # plot_data: data.frame; data to be plotted with ggplot
  # x: character; name of column in plot_data for x-axis
  # x_lab: character; label for x-axis
  # y: character; name of column in plot_data for y-axis
  # y_lab: character; label for y-axis
  # in_vivo: logical; whether both annotations were generated in an in vivo setting
  feature_cor <- round(as.numeric(cor(plot_data[x], plot_data[y], 
                                      method="spearman")), 
                       digits=2)
  cor_plot <- ggplot(plot_data, aes_string(x=x, y=y)) + 
    geom_point(col=ifelse(in_vivo, "black", "darkgrey")) +
    geom_smooth(method=lm, formula=y~x) + theme_bw() + 
    xlab(x_lab) + ylab(y_lab) +
    geom_label(data=data.frame(x=-Inf, y=Inf, label=paste("cor =", feature_cor)),
               aes(x=x, y=y, label=label),
               hjust=0, vjust=1, col="red")
  return(cor_plot)
}

ggplot_violin <- function(plot_data, x_var, y_var="Estimate", fill_var=NULL,
                          x_lab="", y_lab="rate above no activator (RFU/min)",
                          binwidth=1) {
  # plot_data: data.frame ; input to ggplot()
  # x_var: character; x argument for aes_string()
  # y_var: character; y argument for aes_string()
  # fill_var: character; fill argument for aes_string
  # xlab: character; argument for xlab()
  # ylab: character; argument for ylab()
  # binwidth: numeric; argument for geom_dotplot
  if(is.null(fill_var)) {
    fill_var <- x_var
  }
  ggplot(plot_data, aes_string(x=x_var, y=y_var)) + 
    geom_violin(aes_string(fill=fill_var), alpha=0.25, 
                draw_quantiles=c(0.25, 0.75), linetype="dashed") +
    geom_violin(fill="transparent", draw_quantiles=0.5) +
    geom_dotplot(binaxis="y", stackdir="center", binwidth=binwidth) + 
    theme_bw() + xlab(x_lab) + ylab(y_lab)
}

compute_nt_freq <- function(sequences, bases=c("A", "U", "C", "G")) {
  # sequences: character vector; sequences of same length
  # bases: character vector; letters to compute frequences of at each position
  seq_length <- unique(nchar(sequences))
  seq_nt <- matrix(unlist(strsplit(sequences, split="")), ncol=seq_length, byrow=T)
  seq_freq <- sapply(seq(seq_length),
                     function(x) {
                       sapply(bases,
                              function(base) {
                                mean(seq_nt[,x]==base)
                              })
                     })
  rownames(seq_freq) <- bases
  return(data.frame(t(seq_freq)))
}

calculate_relative_info <- function(q_ik, p_ik) {
  # q_ik: data.frame; base frequencies per position in queries (bases as columns)
  # p_ik: data.frame; base frequencies per position in background (bases as columns)
  num_positions <- nrow(q_ik)
  if(nrow(p_ik) != num_positions) {
    stop("q_ik and p_ik have differing number of positions")
  }
  if(any(!(colnames(q_ik) %in% colnames(p_ik)))) {
    stop("q_ik and p_ik have differing alphabets")
  }
  info_content <- sapply(seq(num_positions),
                         function(x) {
                           sum(q_ik[x,] * log2(q_ik[x,] / p_ik[x,]))
                         })
  denominator <- sapply(seq(num_positions),
                        function(x) {
                          sum(q_ik[x,] / p_ik[x,])
                        })
  base_heights <- sapply(colnames(q_ik),
                         function(base) {
                           q_ik[[base]]/p_ik[[base]]/denominator*info_content
                         })
  return(t(base_heights))
}

project_dir <- file.path(here(), "NCR")
ref_dir <- file.path(here(), "ref_data")
mixed <- T
```

# Data
* 204 guides (192 20mers, 12 28mers) screened in triplicate at 0 and 100 fM activator
    + **96 "good" guides (NCR_504-599)**: selected for good secondary structure (maintains crRNA hairpin, no folding in spacer) and high sensitivity to SARS-CoV-2 genomic diversity; varies by whether target is structured (predicted by SHAPE-MaP and DMA-MaPseq) and whether guide targets any organism in NCBI viral and bacterial database
    + **15 Ott lab guides (NCR_600-614)**
    + **48 control "bad" guides (NCR_1305-1352)**: selected for bad secondary structure (crRNA folding interferes with hairpin structure and/or significant folding in spacer)
    + **48 random guides tiling SARS-CoV-2 (NCR_1353-1400)**: good secondary structure, high sensitivity
    + **12 28mer guides (NCR_1407-1418)**: span range of predicted activity by ADAPT (also 28mer)
* plate controls: 612_Control and No_Protein
* repeat sequence: uagaccaccccaaaaaugaaggggacuaaaac

Example structure of a "good" guide

```{r good_guide}
knitr::include_graphics(path.expand(file.path(project_dir, "good_guide.png")))
```

```{r load_guide_features}
order1 <- read.delim(file.path(project_dir, "primary_guide_design_20200710.tsv"))
order2 <- read.delim(file.path(project_dir, "primary_control_guides_20210129.tsv"))
order2$spacer <- substr(order2$sequence, 33, 52)
adapt_28mer <- read.delim(file.path(project_dir, "adapt_28mers.tsv"))
guide_28mer <- read.delim(file.path(project_dir, "cas13a_28nt_results_summary.txt"))
guide_28mer <- guide_28mer[match(adapt_28mer$target.sequences, guide_28mer$target),]
guide_28mer$NCR.id <- adapt_28mer$New.Name
guide_features <- intersect(intersect(colnames(order1), colnames(order2)),  colnames(guide_28mer))
guide_features <- rbind(order1[, guide_features], order2[, guide_features], guide_28mer[, guide_features])
guide_features$crRNA_spacer_basepairs <- factor(guide_features$crRNA_spacer_basepairs, 
                                                levels=sort(unique(guide_features$crRNA_spacer_basepairs)))
guide_features$group <- sapply(seq(nrow(guide_features)),
                               function(x) {
                                 if(guide_features$has_crRNA_hairpin[x] & 
                                    guide_features$crRNA_spacer_basepairs[x] == 0) {
                                   return("good guide")
                                 } else {
                                   paste0(ifelse(guide_features$has_crRNA_hairpin[x], 
                                                 "hasHairpin", "noHairpin"), "/",
                                          guide_features$crRNA_spacer_basepairs[x], "bp")
                                 }
                               })
guide_features$group <- factor(guide_features$group,
                               levels=c("good guide", 
                                        "hasHairpin/4bp", "noHairpin/4bp",
                                        "hasHairpin/6bp",
                                        "noHairpin/7bp",
                                        "noHairpin/8bp",
                                        "noHairpin/9bp",
                                        "hasHairpin/10bp", "noHairpin/10bp",
                                        "noHairpin/11bp",
                                        "hasHairpin/12bp",
                                        "noHairpin/13bp",
                                        "noHairpin/14bp",
                                        "noHairpin/15bp",
                                        "hasHairpin/16bp", "noHairpin/16bp",
                                        "noHairpin/17bp"))
guide_features$class <- sapply(as.numeric(sub("NCR_", "", guide_features$NCR.id)),
                               function(x) {
                                 if(x >= 504 & x <= 599) {
                                   return("good")
                                 } else {
                                   if(x >= 600 & x <= 614) {
                                     return("Ott")
                                   } else {
                                     if(x >= 1305 & x <= 1352) {
                                       return("bad")
                                     } else {
                                       if(x >= 1353 & x <= 1400) {
                                         return("random")
                                       } else {
                                         if(x >= 1407 & x <= 1418) {
                                           return("28mer")
                                         } else {
                                           return(NA)
                                         }
                                       }
                                     }
                                   }
                                 }
                               })
guide_features$class <- factor(guide_features$class, levels=c("good", "Ott", "bad", "random", "28mer"))
```

```{r guide_features_crRNA_structure}
RNAfold_20nt <- read.table(file.path(here(), "outputs", "cas13a_20nt", 
                                     "score_RNAfold_crRNAs.txt"), header=T)
RNAfold_28nt <- read.table(file.path(here(), "outputs", "cas13a_28nt",
                                     "score_RNAfold_crRNAs.txt"), header=T)
guide_features$structure <- sapply(seq(nrow(guide_features)),
                                   function(x) {
                                     tmp_start <- guide_features$start[x]
                                     tmp_spacer_length <- nchar(guide_features$spacer[x])
                                     tmp_structure <- subset(get(paste0("RNAfold_", 
                                                                        tmp_spacer_length, 
                                                                        "nt")),
                                                             start == tmp_start)
                                     return(tmp_structure$structure[1])
                                   })
```

## Viral secondary structure

* Manfredonia et al. *NAR* (2020)
    + SHAPE-MaP/DMS-MaPseq of *in vitro* refolded viral genome in ~500bp tiles
    + data: sequences of single-stranded (low Shannon entropy, high SHAPE) regions (3599 nt; 12.0% of genome)
    + per-guide metric: % of target labeled as single-stranded
* Lan et al. *bioRxiv* (2020)
    + DMS-MaPseq in infected Vero E6 cells
    + data: coordinates of unstructured/structured regions (6010 nt; 20.1% of genome)
    + per-guide metric: % of target labeled as unstructured
* Sun et al. *Cell* (2021)
    + icSHAPE in infected Huh7.5.1 cells & on extracted RNA
    + data: icSHAPE score [0,1] per nucleotide
    + per-guide metric: mean icSHAPE score (higher = more likely to be single-stranded)
* Huston et al. *Molecular Cell* (2021)
    + SHAPE-MaP of purified RNA from infected Vero E6 cells
    + data: RNAstructure connectivity table (CT) per nucleotide (12527 positions unbound; 41.9% of genome)
    + per-guide metric: % target labeled as unpaired

```{r load_viral_genome_structure, fig.height=4}
viral_genome <- Biostrings::readDNAStringSet(file.path(ref_dir, "NC_045512v2.fa"))
viral_genome <- data.frame(pos = seq(nchar(viral_genome)),
                           base = unlist(strsplit(as.character(viral_genome), 
                                                  split="")),
                           spacer_base = unlist(strsplit(as.character(complement(viral_genome)), 
                                                         split="")),
                           row.names=NULL)

# Manfredonia et al. NAR (2020)
manfredonia_prefix <- "manfredonia_2020_lowShannon_highSHAPE"
manfredonia_sam_fname <- file.path(ref_dir, paste0(manfredonia_prefix, ".sam"))
if(!file.exists(manfredonia_sam_fname)) {
  manfredonia <- openxlsx::read.xlsx(file.path(ref_dir, paste0(manfredonia_prefix, ".xlsx")),
                                   colNames=T, startRow=2)
  manfredonia_fasta <- Biostrings::DNAStringSet(manfredonia$Sequence)
  names(manfredonia_fasta) <- with(manfredonia, paste("region", From, To, sep="_"))
  manfredonia_fasta_fname <- file.path(ref_dir, paste0(manfredonia_prefix, ".fa"))
  Biostrings::writeXStringSet(manfredonia_fasta, filepath=manfredonia_fasta_fname)
  system(paste("bowtie --norc -v 2 -S -f", file.path(ref_dir, "wuhCor1"),
               manfredonia_fasta_fname, ">", manfredonia_sam_fname))
}
## add annotation
manfredonia_sam <- load_sam(manfredonia_sam_fname)
viral_genome$manfredonia <- "."
for(x in seq(nrow(manfredonia_sam))) {
  region_start <- manfredonia_sam$pos[x]
  region_end <- region_start + nchar(manfredonia_sam$seq[x]) - 1
  viral_genome$manfredonia[region_start:region_end] <- "lowShannon/highSHAPE"
}

# Lan et al. bioRxiv (2020)
lan_structured_prefix <- "lan_2020_structured"
lan_structured_sam_fname <- file.path(ref_dir, paste0(lan_structured_prefix, ".sam"))
if(!file.exists(lan_structured_sam_fname)) {
  lan_structured <- read.csv(file.path(ref_dir, paste0(lan_structured_prefix, ".csv")))
  lan_structured <- cbind("MN985325.1", lan_structured)
  lan_structured_bed_fname <- file.path(ref_dir, paste0(lan_structured_prefix, ".bed"))
  write.table(lan_structured, file=lan_structured_bed_fname,
              sep="\t", quote=F, row.names=F, col.names=F)
  lan_structured_fasta_fname <- file.path(ref_dir, paste0(lan_structured_prefix, ".fa"))
  system(paste("bedtools getfasta -fi",
               file.path(ref_dir, "MN985325v1.fa"),
               "-bed", lan_structured_bed_fname,
               "-fo", lan_structured_fasta_fname))
  system(paste("bowtie --norc -v 2 -S -f", file.path(ref_dir, "wuhCor1"),
               lan_structured_fasta_fname, ">", lan_structured_sam_fname))
}
lan_unstructured_prefix <- "lan_2020_unstructured"
lan_unstructured_sam_fname <- file.path(ref_dir, paste0(lan_unstructured_prefix, ".sam"))
if(!file.exists(lan_unstructured_sam_fname)) {
  lan_unstructured <- read.csv(file.path(ref_dir, paste0(lan_unstructured_prefix, ".csv")))
  lan_unstructured <- cbind("MN985325.1", lan_unstructured)
  lan_unstructured_bed_fname <- file.path(ref_dir, paste0(lan_unstructured_prefix, ".bed"))
  write.table(lan_unstructured, file=lan_unstructured_bed_fname,
              sep="\t", quote=F, row.names=F, col.names=F)
  lan_unstructured_fasta_fname <- file.path(ref_dir, paste0(lan_unstructured_prefix, ".fa"))
  system(paste("bedtools getfasta -fi",
               file.path(ref_dir, "MN985325v1.fa"),
               "-bed", lan_unstructured_bed_fname,
               "-fo", lan_unstructured_fasta_fname))
  system(paste("bowtie --norc -v 2 -S -f", file.path(ref_dir, "wuhCor1"),
               lan_unstructured_fasta_fname, ">", lan_unstructured_sam_fname))
}
# add annotations
lan_structured_sam <- load_sam(lan_structured_sam_fname)
lan_unstructured_sam <- load_sam(lan_unstructured_sam_fname)
viral_genome$lan <- "."
for(x in seq(nrow(lan_structured_sam))) {
  region_start <- lan_structured_sam$pos[x]
  region_end <- region_start + nchar(lan_structured_sam$seq[x]) - 1
  viral_genome$lan[region_start:region_end] <- "structured"
}
for(x in seq(nrow(lan_unstructured_sam))) {
  region_start <- lan_unstructured_sam$pos[x]
  region_end <- region_start + nchar(lan_unstructured_sam$seq[x]) - 1
  region <- viral_genome$lan[region_start:region_end]
  region <- paste0(region, "_unstructured")
  viral_genome$lan[region_start:region_end] <- region
}
viral_genome$lan[viral_genome$lan=="._unstructured"] <- "unstructured"
viral_genome$lan[viral_genome$lan=="structured_unstructured"] <- "both"

# Sun et al. Cell (2021)
sun_invivo <- openxlsx::read.xlsx(file.path(ref_dir, "sun_2021_icSHAPE.xlsx"),
                                  sheet="SARS2-invivo", na.strings="NULL")
viral_genome$sun_invivo <- sun_invivo$`icSHAPE-score`
sun_invitro <- openxlsx::read.xlsx(file.path(ref_dir, "sun_2021_icSHAPE.xlsx"),
                                   sheet="SARS2-invitro", na.strings="NULL")
viral_genome$sun_invitro <- sun_invitro$`icSHAPE-score`

# Huston et al. Molecular Cell*(2021)
huston <- read.table(file.path(ref_dir, "huston_2021_structure.ct"), skip=1,
                     col.names=c("index", "base", "previous_index", "next_index", 
                                 "paired_to", "natural_numbering"))
viral_genome$huston <- as.numeric(huston$paired_to)

(ggplot(viral_genome, aes(x=manfredonia)) + geom_bar() + theme_bw() + 
    ggtitle("Manfredonia") + xlab("") + ylab("# genomic positions") +
    theme(axis.text.x=element_text(angle=45, hjust=1))) +
  (ggplot(viral_genome, aes(x=lan)) + geom_bar() + theme_bw() +
     ggtitle("Lan") + xlab("") + ylab("# genomic positions") + 
     theme(axis.text.x=element_text(angle=45, hjust=1))) +
  (ggplot(viral_genome, aes(x=sun_invivo)) + geom_density(fill="grey") + theme_bw() +
     ggtitle("Sun (in vivo)") + xlab("") + ylab("") +
     theme(axis.text.x=element_text(angle=45, hjust=1))) +
  (ggplot(viral_genome, aes(x=sun_invitro)) + geom_density(fill="grey") + theme_bw() +
     ggtitle("Sun (in vitro)") + xlab("") + ylab("") +
     theme(axis.text.x=element_text(angle=45, hjust=1))) +
  (ggplot(viral_genome, aes(x=(huston==0))) + geom_bar() + theme_bw() +
     ggtitle("Huston") + xlab("") + ylab("# genomic positions")) + 
  plot_layout(nrow=1, widths=rep(1, 5))
```

### Correlation between annotations: over entire spacer target

```{r guide_viral_structure, fig.height=8}
guide_structure <- sapply(seq(nrow(guide_features)),
                          function(x) {
                            region_start <- guide_features$start[x]
                            region_end <- region_start + nchar(guide_features$spacer[x]) - 1
                            region_features <- viral_genome[match(region_start:region_end,
                                                                  viral_genome$pos),]
                            region_annotations <- c(manfredonia = mean(region_features$manfredonia == "lowShannon/highSHAPE"),
                                                    lan = mean(region_features$lan == "unstructured"),
                                                    sun_invivo = mean(region_features$sun_invivo),
                                                    sun_invitro = mean(region_features$sun_invitro),
                                                    huston = mean(region_features$huston == 0))
                            return(region_annotations)
                          })
guide_structure <- data.frame(t(guide_structure))
guide_features <- cbind(guide_features, guide_structure)

manfredonia_lan <- plot_guide_feature_cor(guide_structure,
                                          "manfredonia", "Manfredonia",
                                          "lan", "Lan")
manfredonia_sun_invivo <- plot_guide_feature_cor(guide_structure,
                                                 "manfredonia", "Manfredonia",
                                                 "sun_invivo", "Sun (in vivo)")
manfredonia_sun_invitro <- plot_guide_feature_cor(guide_structure,
                                                  "manfredonia", "Manfredonia",
                                                  "sun_invitro", "Sun (in vitro)")
manfredonia_huston <- plot_guide_feature_cor(guide_structure,
                                             "manfredonia", "Manfredonia",
                                             "huston", "Huston")
lan_sun_invivo <- plot_guide_feature_cor(guide_structure,
                                         "lan", "Lan",
                                         "sun_invivo", "Sun (in vivo)",
                                         in_vivo=T)
lan_sun_invitro <- plot_guide_feature_cor(guide_structure,
                                          "lan", "Lan",
                                          "sun_invitro", "Sun (in vitro)")
lan_huston <- plot_guide_feature_cor(guide_structure,
                                     "lan", "Lan",
                                     "huston", "Huston",
                                     in_vivo=T)
sun_invivo_sun_invitro <- plot_guide_feature_cor(guide_structure,
                                                 "sun_invivo", "Sun (in vivo)",
                                                 "sun_invitro", "Sun (in vitro)")
sun_invivo_huston <- plot_guide_feature_cor(guide_structure,
                                            "sun_invivo", "Sun (in vivo)",
                                            "huston", "Huston",
                                            in_vivo=T)
sun_invitro_huston <- plot_guide_feature_cor(guide_structure,
                                             "sun_invitro", "Sun (in vitro)",
                                             "huston", "Huston")

manfredonia_huston + lan_huston + sun_invivo_huston + sun_invitro_huston +
  manfredonia_sun_invitro + lan_sun_invitro + sun_invivo_sun_invitro + plot_spacer() +
  manfredonia_sun_invivo + lan_sun_invivo + plot_spacer() + plot_spacer() +
  manfredonia_lan + plot_spacer() + plot_spacer() + plot_spacer()
```

* Annotations from Manffedonia (2020) correlate poorly with other annotations
    + Only 12% of genome is labeled
    + Experimental data generated on 500bp tiling of refolded viral genome
* Relatively high correlation between annotations generated from *in vivo* settings
    + Lan (2020)
    + Sun (2021)
    + Huston (2021)

### Intersection of *in vivo* viral structure predictions

```{r in_vivo_intersection, fig.height=4}
guide_features$lan_sun_huston <- sapply(seq(nrow(guide_features)),
                                        function(x) {
                                          tmp_sun <- guide_features$sun_invivo[x]
                                          tmp_huston <- guide_features$huston[x]
                                          tmp_lan <- guide_features$lan[x]
                                          if(tmp_sun > 0.25 & tmp_huston >= 0.75) {
                                            return("unstructured")
                                          } else {
                                            if(tmp_sun <= 0.25 & tmp_huston <= 0.25) {
                                              return("structured")
                                            } else {
                                              return("intermediate")
                                            }
                                          }
                                        })
guide_features$lan_sun_huston <- factor(guide_features$lan_sun_huston,
                                        levels=c("unstructured", "intermediate", "structured"))

ggplot(guide_features, aes(x=sun_invivo, y=huston, col=lan, shape=lan_sun_huston)) +
  geom_point() + theme_bw() + labs(col="Lan", shape="Structure (intersection)") + 
  xlab("Sun (in vivo)") + ylab("Huston")
```

### Correlation between annotations: over spacer seed region (positions 8-12 on spacer)

```{r guide_viral_structure_seed, fig.height=8}
guide_structure_seed <- sapply(seq(nrow(guide_features)),
                          function(x) {
                            region_start <- guide_features$start[x] + 7
                            region_end <- guide_features$start[x] + 11
                            region_features <- viral_genome[match(region_start:region_end,
                                                                  viral_genome$pos),]
                            region_annotations <- c(manfredonia = mean(region_features$manfredonia == "lowShannon/highSHAPE"),
                                                    lan = mean(region_features$lan == "unstructured"),
                                                    sun_invivo = mean(region_features$sun_invivo),
                                                    sun_invitro = mean(region_features$sun_invitro),
                                                    huston = mean(region_features$huston == 0))
                            return(region_annotations)
                          })
guide_structure_seed <- data.frame(t(guide_structure_seed))
colnames(guide_structure_seed) <- paste0(colnames(guide_structure_seed), "_seed")
guide_features <- cbind(guide_features, guide_structure_seed)

manfredonia_lan_seed <- plot_guide_feature_cor(guide_structure_seed,
                                          "manfredonia_seed", "Manfredonia",
                                          "lan_seed", "Lan")
manfredonia_sun_invivo_seed <- plot_guide_feature_cor(guide_structure_seed,
                                                 "manfredonia_seed", "Manfredonia",
                                                 "sun_invivo_seed", "Sun (in vivo)")
manfredonia_sun_invitro_seed <- plot_guide_feature_cor(guide_structure_seed,
                                                  "manfredonia_seed", "Manfredonia",
                                                  "sun_invitro_seed", "Sun (in vitro)")
manfredonia_huston_seed <- plot_guide_feature_cor(guide_structure_seed,
                                             "manfredonia_seed", "Manfredonia",
                                             "huston_seed", "Huston")
lan_sun_invivo_seed <- plot_guide_feature_cor(guide_structure_seed,
                                         "lan_seed", "Lan",
                                         "sun_invivo_seed", "Sun (in vivo)",
                                         in_vivo=T)
lan_sun_invitro_seed <- plot_guide_feature_cor(guide_structure_seed,
                                          "lan_seed", "Lan",
                                          "sun_invitro_seed", "Sun (in vitro)")
lan_huston_seed <- plot_guide_feature_cor(guide_structure_seed,
                                     "lan_seed", "Lan",
                                     "huston_seed", "Huston",
                                     in_vivo=T)
sun_invivo_sun_invitro_seed <- plot_guide_feature_cor(guide_structure_seed,
                                                 "sun_invivo_seed", "Sun (in vivo)",
                                                 "sun_invitro_seed", "Sun (in vitro)")
sun_invivo_huston_seed <- plot_guide_feature_cor(guide_structure_seed,
                                            "sun_invivo_seed", "Sun (in vivo)",
                                            "huston_seed", "Huston",
                                            in_vivo=T)
sun_invitro_huston_seed <- plot_guide_feature_cor(guide_structure_seed,
                                             "sun_invitro_seed", "Sun (in vitro)",
                                             "huston_seed", "Huston")

manfredonia_huston_seed + lan_huston_seed + sun_invivo_huston_seed + sun_invitro_huston_seed +
  manfredonia_sun_invitro_seed + lan_sun_invitro_seed + sun_invivo_sun_invitro_seed + plot_spacer() +
  manfredonia_sun_invivo_seed + lan_sun_invivo_seed + plot_spacer() + plot_spacer() +
  manfredonia_lan_seed + plot_spacer() + plot_spacer() + plot_spacer()
```

```{r in_vivo_kmeans}
set.seed(10)
invivo_kmeans_2 <- kmeans(guide_features[, c("lan", "sun_invivo", "huston")], 2)
invivo_kmeans_2_labels <- rank(rowMeans(invivo_kmeans_2$centers))
guide_features$invivo_kmeans_2 <- invivo_kmeans_2$cluster
guide_features$invivo_kmeans_2 <- as.factor(guide_features$invivo_kmeans_2)
levels(guide_features$invivo_kmeans_2) <- c("structured", "unstructured")[invivo_kmeans_2_labels]

set.seed(20)
invivo_kmeans_3 <- kmeans(guide_features[, c("lan", "sun_invivo", "huston")], 3)
invivo_kmeans_3_labels <- rank(rowMeans(invivo_kmeans_3$centers))
guide_features$invivo_kmeans_3 <- invivo_kmeans_3$cluster
guide_features$invivo_kmeans_3 <- as.factor(guide_features$invivo_kmeans_3)
levels(guide_features$invivo_kmeans_3) <- c("structured", "intermediate", "unstructured")[invivo_kmeans_3_labels]
```

```{r map_96_to_384}
mapping_384_to_96 <- data.frame(matrix(NA, nrow=16, ncol=24))
rownames(mapping_384_to_96) <- LETTERS[1:16]
for(x in 1:8) {
  for(y in 1:12) {
    tmp_rows <- c(2*x-1, 2*x-1, 2*x)
    tmp_cols <- c(2*y-1, 2*y, 2*y)
    tmp_label <- paste0(LETTERS[x], y)
    for(z in 1:3) {
      mapping_384_to_96[tmp_rows[z], tmp_cols[z]] <- tmp_label
    }
  }
}
mapping_384_to_96 <- data.frame(well_96 = unlist(mapping_384_to_96),
                                well_384 = paste0(rep(LETTERS[1:16], times=24),
                                                  rep(1:24, each=16)))
mapping_384_to_96[is.na(mapping_384_to_96)] <- "empty"
```

```{r load_platemaps}
expt_names <- c("GS1_1", "GS1_2", "GS2_1", "GS2_2", "GS3")
plate_maps <- xlsx::read.xlsx(file.path(project_dir, "AC_Guide_Screening.xlsx"),
                              sheetName="Plate Setups")
plate_maps <- subset(plate_maps, NA. %in% LETTERS[1:8])[, 3:14]
plate_maps <- split(plate_maps, rep(expt_names, each=8))
plate_maps <- lapply(seq_along(plate_maps),
                     function(x) {
                       if(x %in% 1:4) {
                         data.frame(sample = paste0(unlist(plate_maps[[x]]), 
                                                    rep(c("_noActivator", "_100fM"), each=48)),
                                    well_96 = paste0(rep(LETTERS[1:8], times=12),
                                                     rep(1:12, each=8)))
                       } else {
                         tmp_map <- data.frame(sample = paste0(unlist(plate_maps[[x]][1:4,]),
                                                               c(rep(rep(c("_noActivator", "_100fM"), 
                                                                         each=2),
                                                                     times=12))),
                                               well_96 = paste0(rep(LETTERS[1:4], times=12),
                                                                rep(1:12, each=4)))
                         return(subset(tmp_map, !grepl("NA", tmp_map$sample)))
                       }
                     })
names(plate_maps) <- expt_names
```

```{r load_plate_data}
data_dir <- file.path(project_dir, "100fM_data")
data_fnames <- list.files(data_dir)
plate_data <- lapply(data_fnames,
                     function(x) {
                       tmp_plate <- sub("_100fM_Plate", "", 
                                        sub("-", "_", 
                                            sub(".xlsx", "", x)))
                       row_indices <- 57:117
                       tmp_data <- xlsx::read.xlsx(file.path(data_dir, x), 
                                                   sheetIndex=1, rowIndex=row_indices, header=T)
                       tmp_data <- lapply(seq(nrow(tmp_data)),
                                          function(x) {
                                            data.frame(tmp_data[x, 1:3],
                                                       unlist(tmp_data[x, 4:387]),
                                                       colnames(tmp_data)[4:387],
                                                       row.names = NULL)
                                          })
                       tmp_data <- do.call(rbind, tmp_data)
                       colnames(tmp_data) <- c("cycle_number", "time", "temp", "signal", "well_384")
                       tmp_data$well_96 <- mapping_384_to_96$well_96[match(tmp_data$well_384,
                                                                           mapping_384_to_96$well_384)]
                       tmp_platemap <- plate_maps[[tmp_plate]]
                       tmp_data$sample <- tmp_platemap$sample[match(tmp_data$well_96,
                                                                    tmp_platemap$well_96)]
                       tmp_data$plate <- tmp_plate
                       return(tmp_data)
                     })
plate_data <- do.call(rbind, plate_data)
plate_data$guide_id <- sapply(plate_data$sample,
                              function(x) {
                                ifelse(grepl("_", x),
                                       sub(" ", "_", sub("_.*", "", x)),
                                       "empty")
                              })

plate_data$activator <- factor(sapply(plate_data$sample,
                                      function(x) {
                                        ifelse(grepl("_", x),
                                               sub(".*_", "", x),
                                               "empty")
                                      }), 
                               levels=c("100fM", "noActivator", "empty"))
plate_data$cycle_number <- as.numeric(plate_data$cycle_number)
plate_data$signal <- as.numeric(plate_data$signal)
plate_data$time <- plate_data$time/60 # convert from seconds to minutes
```

```{r background_subtract}
## background subtract: empty wells
background_empty <- mean(subset(plate_data, is.na(sample))$signal, na.rm=T)
plate_data$signal_bkgdSubtract <- plate_data$signal - background_empty
plate_data <- subset(plate_data, !is.na(sample))
plate_data$activator <- droplevels(plate_data$activator)
```

# Exploratory data analysis: controls

## Plate controls: 612_Control and No_Protein

```{r analyze_plate_controls}
## plot controls: No_Protein and 612_Control
control_data <- subset(plate_data, 
                       guide_id %in% c("No_Protein", "612_Control"))
control_outliers <- rbind(subset(control_data,
                                 cycle_number==60 & 
                                   guide_id=="612_Control" & 
                                   signal_bkgdSubtract > 15000),
                          subset(control_data,
                                 cycle_number==60 &
                                   guide_id=="612_Control" &
                                   activator=="noActivator" &
                                   signal_bkgdSubtract > 4500),
                          subset(control_data,
                                 cycle_number==60 &
                                   guide_id=="No_Protein" &
                                   signal_bkgdSubtract > 3500))
control_data$group <- with(control_data, paste(plate, well_384, sep="_"))
ggplot(control_data, aes(x=time, y=signal_bkgdSubtract, col=plate)) +
  theme_bw() + xlab("time (min)") + ylab("background-subtracted fluorescence (RFU)") +
  geom_line(aes(group=group), alpha=0.5) + facet_grid(guide_id ~ activator) +
  geom_text(data=control_outliers, 
            aes(x=time, y=signal_bkgdSubtract, label=well_384),
            nudge_x=-5, col="black", size=2)
```

* Plate GS2-2 outliers: O 9-12, 21-24
* No_Protein (+/- activator) and 612_Control (- activator) controls cluster tightly across plates
* Variable spread of 612_Control activity with 100 fM activator

## No activator controls

```{r analyze_noActivator, fig.height=7}
noActivator_data <- subset(plate_data, activator=="noActivator")
noActivator_outliers <- rbind(subset(noActivator_data,
                                     cycle_number==60 &
                                       plate=="GS1_2" &
                                       signal_bkgdSubtract > 9000),
                              subset(noActivator_data,
                                     cycle_number==60 &
                                       plate=="GS2_1" &
                                       signal_bkgdSubtract > 10000),
                              subset(noActivator_data,
                                     plate=="GS2_1" &
                                       cycle_number==23 &
                                       signal_bkgdSubtract > 60000))
# calculate rates per guide
noActivator_rates <- split(noActivator_data, noActivator_data$guide_id)
noActivator_rates <- sapply(noActivator_rates, compute_rate)
noActivator_rates <- data.frame(guide_id=sub(".time", "", names(noActivator_rates)),
                                rate=noActivator_rates,
                                row.names=NULL)
noActivator_rates$plate <- noActivator_data$plate[match(noActivator_rates$guide_id,
                                                        noActivator_data$guide_id)]
# plot
(ggplot(noActivator_data, aes(x=time, y=signal_bkgdSubtract, col=plate)) +
    theme_bw() + xlab("time (min)") + ylab("background-subtracted fluoresence (RFU)") +
    geom_line(aes(group=well_384), alpha=0.5) + facet_grid(~plate) +
    geom_text(data=noActivator_outliers,
              aes(x=time, y=signal_bkgdSubtract, label=well_384), col="black") +
    theme(legend.position="none")) /
  (ggplot(noActivator_rates, aes(x=plate, y=rate, fill=plate)) +
     geom_violin() + theme_bw() + xlab("") + theme(legend.position="none") +
     ylim(-5, 30) + ggtitle("", subtitle="removing outliers")) +
  plot_layout(heights=c(2,1))
```

* No activator varies by plate (and presumably by guide) 
+ Want to be able to account for additional gain in rate above no activator background
* Outliers on plate GS2-1 all correspond to NCR_1344

```{r NCR_1344_structure}
knitr::include_graphics(path.expand(file.path(project_dir, "NCR_1344.png")))
```

# Data analysis

## Methods:

* Mixed linear regression to compute rate per condition across triplicates:
    + per guide: $\text{signal}_i(t) = (\beta_0 + \beta_{0,i}) + (\beta_{\text{100fM}}\cdot\mathbb{I}[\text{100fM}]) \ + (\beta_{t} + \beta_{t,i}) \cdot t \ + (\beta_{t:\text{100fM}}\cdot t\cdot\mathbb{I}[\text{100fM}])$
    + no activator: $\text{signal}_i(t) = (\beta_0 + \beta_{0,i}) +  (\beta_t + \beta_{t,i}) \cdot t_i$
    + 100fM activator: $\text{signal}_i(t) = (\beta_0 + \beta_{0,i} + \beta_{\text{100fM}}) + (\beta_t + \beta_{t,i} +  \beta_{t:\text{100fM}}) \cdot t_i$
    + 100fM rate above no activator: $\beta_{t:\text{100fM}}$
* Background subtract mean of empty wells
* Exclude timepoints before 1000 seconds (first 10 timepoints show dip in some no activator controls)
* Random effects (per 384-well, across timepoints): intercept, slope (wrt time)
* Fixed effects: intercepts ($\beta_0$, $\beta_{\text{100fM}}$), slopes ($\beta_t$, $\beta_{t:\text{100fM}}$)

```{r analyze_data}
guide_ids <- unique(plate_data$guide_id)
guide_ids <- guide_ids[!(guide_ids %in% c("612_Control", "No_Protein"))]
all_results <- lapply(guide_ids,
                      function(x) {
                        analyze_guide(plate_data, guide=x, time_start=15,
                                      signal="signal_bkgdSubtract", 
                                      mixed_model=mixed)
                      })
names(all_results) <- guide_ids
all_plots <- lapply(seq_along(all_results),
                    function(x) {
                      guide_id <- paste0("NCR_", names(all_results)[x])
                      which_row <- match(guide_id, guide_features$NCR.id)
                      has_hairpin <- guide_features$has_crRNA_hairpin[which_row]
                      spacer_bp <- guide_features$crRNA_spacer_basepairs[which_row]
                      plot_subtitle <- paste(ifelse(has_hairpin, "has hairpin", "no hairpin"), "|",
                                             spacer_bp, "basepaired positions in spacer")
                      all_results[[x]]$plot + ggtitle(guide_id, subtitle=plot_subtitle)
                    })
names(all_plots) <- guide_ids
plots_by_group <- split(all_plots, 
                        guide_features$group[match(paste0("NCR_", names(all_plots)),
                                                   guide_features$NCR.id)])
```

### Example traces:

* dots: signal per 384-well per time
* solid lines: regression fit

```{r example_traces, fig.height=8}
all_plots[["522"]] / all_plots[["527"]]
```

## Results: 100 fM rates (above no activator)

### Summary of screen

```{r NCR_1324_structure, fig.height=8, include=F}
NCR_1324_noOutlier <- subset(plate_data, guide_id=="1324")
NCR_1324_outlierWell <- NCR_1324_noOutlier$well_384[which.max(NCR_1324_noOutlier$signal_bkgdSubtract)]
NCR_1324_noOutlier <- subset(NCR_1324_noOutlier, well_384 != NCR_1324_outlierWell)
NCR_1324_noOutlier_results <- analyze_guide(NCR_1324_noOutlier, guide="1324", mixed_model=mixed)
all_plots[["1324"]] / (NCR_1324_noOutlier_results$plot + ggtitle("NCR_1324", subtitle="remove outlier"))
all_results[["1324"]] <- NCR_1324_noOutlier_results
```

```{r aggregate_coefficients}
all_coef <- lapply(all_results, function(x) x$coef)
all_coef <- do.call(rbind, all_coef)
all_coef <- cbind(all_coef,
                  guide_features[match(paste0("NCR_", all_coef$guide_id),
                                       guide_features$NCR.id),])
all_coef$plate <- plate_data$plate[match(all_coef$guide_id, plate_data$guide_id)]
```

```{r compute_100fM_rate}
guide_rate_100fM <- aggregate(Estimate ~ guide_id, 
                              data=subset(all_coef, grepl("time", all_coef$coef)),
                              FUN=sum)
guide_rate_100fM <- cbind(guide_rate_100fM,
                          guide_features[match(paste0("NCR_", guide_rate_100fM$guide_id),
                                               guide_features$NCR.id),])
```

```{r plot_rates, fig.height=16}
guide_rate <- subset(all_coef, grepl(":", all_coef$coef))
ggplot(guide_rate, aes(x=Estimate, y=guide_id, col=plate)) + geom_point() +
  geom_errorbar(aes(xmin=Estimate-qnorm(0.05)*Std.Error,
                    xmax=Estimate+qnorm(0.05)*Std.Error)) + 
  theme_bw() + geom_vline(xintercept=0,) + theme(axis.text.y=element_text(size=6)) + 
  xlab("rate above no activator (RFU/min)") + ylab("guide id")

# rotated plot
# ggplot(guide_rate, aes(x=guide_id, y=Estimate, col=plate)) + geom_point() +
#   geom_errorbar(aes(ymin=Estimate-qnorm(0.05)*Std.Error, 
#                     ymax=Estimate+qnorm(0.05)*Std.Error)) + 
#   theme_bw() + geom_hline(yintercept=0) + scale_x_discrete(limits=rev) +
#   theme(axis.text.x=element_text(angle=90, size=6, hjust=1, vjust=0.5)) + 
#   xlab("") + ylab("rate above no activator (RFU/min)")
```

* Similar spread of rates across plates
+ Expect to see poorer rates for plates GS2-1 and GS2-2 (which harbor the "bad" guides)

### Volcano plots of screening rates

```{r plot_rate_volcano}
ggplot(guide_rate, aes(x=Estimate, y=-log10(p.value), col=plate)) + geom_point() +
  theme_bw() + xlab("rate above no activator (RFU/min)") + ylab("-log10(p)") +
  geom_hline(yintercept=-log10(0.05/nrow(guide_rate)), col="red", linetype="dashed")
```

* Guides on plate GS2-2 have similar rates as to other plates, but are much more variable (p-values closer to 1)

### Rates by position along SARS-CoV-2 genome

```{r rate_by_position}
ggplot(guide_rate, aes(x=start, y=Estimate, 
                       col=crRNA_spacer_basepairs, shape=has_crRNA_hairpin)) + 
  geom_point(size=2, alpha=0.8) + theme_bw() + 
  xlab("genomic position") + ylab("rate above no activator (RFU/min)")
```

Structures of the two guides that performed well (rate > 1 above background) but without hairpin structure (NCR_1346, NCR_1351):

```{r NCR_1346_structure}
knitr::include_graphics(path.expand(file.path(project_dir, "NCR_1346.png")))
```

```{r NCR_1351_structure}
knitr::include_graphics(path.expand(file.path(project_dir, "NCR_1351.png")))
```

### Rates across groups of guide secondary structure

```{r rate_by_spacer_structure}
guide_rate$spacer_bp <- as.numeric(as.character(guide_rate$crRNA_spacer_basepairs))
spacer_structure_cor <- with(guide_rate, cor.test(Estimate, spacer_bp))
ggplot(guide_rate, aes(x=spacer_bp, y=Estimate)) + geom_point() + 
  geom_smooth(method=lm, formula=y~x) + theme_bw() +
  xlab("number of basepaired positions in spacer") + 
  ylab("rate above no activator (RFU/min)") +
  annotate(geom="text", x=0.5, y=-15, 
           label=paste("cor =", round(spacer_structure_cor$estimate, digits=3))) +
  annotate(geom="text", x=0.5, y=-20,
           label=paste("p =", round(spacer_structure_cor$p.value, digits=3))) 
```

* Statistically significant (positive?) correlation between guide activity and spacer structure

```{r rate_by_guidegroup}
ggplot_violin(subset(guide_rate, crRNA_spacer_basepairs %in% c(0, 4, 10, 16)),
              x_var="group", fill_var="has_crRNA_hairpin") + 
  theme(axis.text.x=element_text(angle=90, hjust=1))
```

Determination of how much of predicted hairpin structure needs to be maintained:

```{r hairpin_structure, fig.height=22.5}
hairpin_length <- nchar(guide_features$structure[1]) - 20
good_hairpin <- substr(guide_features$structure[1], 1, hairpin_length)

maintain_hairpin <- sapply(24:hairpin_length,
                           function(x) {
                             good_structure <- substr(good_hairpin, 1, x)
                             tmp_structures <- substr(guide_rate$structure, 1, x)
                             return(tmp_structures == good_structure)
                           })
maintain_hairpin <- data.frame(cbind(guide_rate[, c("guide_id", "start", 
                                                    "Estimate", "crRNA_spacer_basepairs")],
                                     maintain_hairpin))
colnames(maintain_hairpin)[grepl("^X", colnames(maintain_hairpin))] <- paste0("first_", 24:hairpin_length)

maintain_hairpin_noOutliers <- subset(maintain_hairpin, !(guide_id %in% c(1346, 1351)))

maintain_hairpin_plots <- lapply(24:hairpin_length,
                                 function(x) {
                                   hairpin_length <- paste0("first_", x)
                                   tmp_wilcox <- wilcox.test(formula(paste("Estimate ~", 
                                                                           hairpin_length)),
                                                             data=maintain_hairpin_noOutliers)
                                   ggplot_violin(subset(maintain_hairpin_noOutliers,
                                                        crRNA_spacer_basepairs %in% c(0, 4, 10, 16)),
                                                 x_var=hairpin_length, fill_var="crRNA_spacer_basepairs",
                                                 x_lab="maintain hairpin") +
                                     ggtitle("", subtitle=paste("first", x, "bases")) +
                                     annotate(geom="text", x=T, y=80, 
                                              label=paste("p =", round(tmp_wilcox$p.value, digits=3)))
                                 })
wrap_plots(maintain_hairpin_plots, ncol=1, byrow=T)
```

```{r bad_hairpin}
bad_hairpin <- subset(maintain_hairpin, !first_24)$guide_id
subset(guide_features, NCR.id %in% paste0("NCR_", bad_hairpin))[, c("NCR.id", "spacer", "structure")]
```

```{r NCR_1313_structure}
knitr::include_graphics(path.expand(file.path(project_dir, "NCR_1313.png")))
```

```{r NCR_1319_structure}
knitr::include_graphics(path.expand(file.path(project_dir, "NCR_1319.png")))
```

```{r NCR_1332_structure}
knitr::include_graphics(path.expand(file.path(project_dir, "NCR_1332.png")))
```

```{r NCR_1335_structure}
knitr::include_graphics(path.expand(file.path(project_dir, "NCR_1335.png")))
```

```{r NCR_1348_structure}
knitr::include_graphics(path.expand(file.path(project_dir, "NCR_1348.png")))
```

```{r NCR_1349_structure}
knitr::include_graphics(path.expand(file.path(project_dir, "NCR_1349.png")))
```

```{r NCR_1351_structure_2}
knitr::include_graphics(path.expand(file.path(project_dir, "NCR_1351.png")))
```

* "Good" guides exhibit similar 100fM rates as "bad" guides

### Rates across guide ordering group

```{r rate_by_class, fig.height=8}
(ggplot_violin(guide_rate, x_var="class") +
   guides(fill=F) + theme(axis.text.x=element_text(angle=90, hjust=1))) / 
  (ggplot_violin(guide_rate_100fM, x_var="class", y_lab="100fM rate (RFU/min)") + 
     guides(fill=F) + theme(axis.text.x=element_text(angle=90, hjust=1)))
```

### Background (no activator) rate by guide secondary structure

```{r noActivator_rates}
noActivator_rate <- subset(all_coef, coef=="time")
noActivator_rate <- subset(noActivator_rate, guide_id != "1344")
ggplot_violin(subset(noActivator_rate, crRNA_spacer_basepairs %in% c(0, 4, 10, 16)),
              x_var="group", fill_var="has_crRNA_hairpin",
              y_lab="no activator rate (RFU/min)", binwidth=0.25) + 
  labs(fill="maintains crRNA hairpin") + 
  theme(axis.text.x=element_text(angle=90, hjust=1))
```

* In guides that maintain crRNA hairpin: correlation between number of basepaired positions in spacer and background no activator rate

```{r rate_by_gc_content}
ggplot(guide_rate, aes(x=GC_content, y=Estimate)) + geom_point() + 
  geom_smooth(method="loess", formula=y~x) + theme_bw() + 
  xlab("GC content (%)") + ylab("rate above no activator (RFU/min)") 
```

* No correlation between spacer GC content and guide activity

## Comparison of 20mers to 28mers

```{r adapt_comparison}
guide_rate_adapt <- rbind(data.frame(adapt_28mer,
                                     rate=guide_rate$Estimate[match(adapt_28mer$Original.Name,
                                                                    paste0("NCR_", guide_rate$guide_id))],
                                     spacer="20mer"),
                          data.frame(adapt_28mer,
                                     rate=guide_rate$Estimate[match(adapt_28mer$New.Name,
                                                                    paste0("NCR_", guide_rate$guide_id))],
                                     spacer="28mer"))
ggplot(guide_rate_adapt, aes(x=guide.expected.activities, y=rate, col=spacer)) + 
  geom_point() + geom_smooth(method="lm", formula=y~x) + theme_bw() +
  xlab("ADAPT prediction") + ylab("rate above no activator (RFU/min)")
```

## Guide activity by viral structure

```{r rate_by_viral_structure}
# Manfredonia (2020)
guide_rate$manfredonia_cat <- cut(guide_rate$manfredonia, 
                                  breaks=c(0, 0.04, 0.96, 1),
                                  include.lowest=T, right=F,
                                  labels=c("0%", "5-95%", "100%"))
manfredonia_anova <- anova(lm(Estimate~manfredonia_cat, data=guide_rate))
manfredonia_plot <- ggplot_violin(guide_rate, x_var="manfredonia_cat",
                                  x_lab="% basepaired positions in viral target") +
  guides(fill=F) + ggtitle("Manfredonia (2020)", subtitle="Persistent single-strandedness")

# Lan (2020)
guide_rate$lan_cat <- cut(guide_rate$lan,
                          breaks=c(0, 0.04, 0.5, 0.96, 1),
                          include.lowest=T, right=F,
                          labels=c("0%", "5-49%", "50-95%", "100%"))
lan_fit <- data.frame(summary(lm(Estimate~lan_cat, data=guide_rate))$coefficients)
lan_wilcox <- with(guide_rate,
                   wilcox.test(Estimate[lan_cat=="0%"],
                               Estimate[lan_cat=="50-95%"]))
lan_plot <- (ggplot(guide_rate, aes(x=lan, y=Estimate)) + 
               geom_point() + geom_smooth(method=loess, formula=y~x) + theme_bw() + 
               xlab("% unstructured positions in viral target") + 
               ylab("rate above no activator (RFU/min)") +
               ggtitle("Lan (2020)", subtitle="Unstructured region")) + 
  (ggplot_violin(guide_rate, x_var="lan_cat",
                 x_lab="% unstructured positions in viral target") + 
     annotate(geom="text", x="50-95%", y=80, 
              label=paste("* p=", round(lan_wilcox$p.value, digits=3))))
lan_spacer_bp_fit <- lm(Estimate~spacer_bp+lan_cat, data=guide_rate)
lan_spacer_bp_coef <- data.frame(summary(lan_spacer_bp_fit)$coefficients)
lan_spacer_bp_coef$coef <- rownames(lan_spacer_bp_coef)
lan_spacer_bp_data <- guide_rate[, c("Estimate", "spacer_bp", "lan", "lan_cat")]
lan_spacer_bp_data$rate <- with(lan_spacer_bp_data,
                                Estimate - spacer_bp*lan_spacer_bp_coef$Estimate[2])
lan_spacer_bp_wilcox <- with(lan_spacer_bp_data,
                             wilcox.test(rate[lan_cat=="0%"],
                                         rate[lan_cat=="50-95%"]))
lan_spacer_bp_plot <- ((ggplot(guide_rate, aes(x=spacer_bp, y=Estimate)) + 
                          geom_point() + 
                          geom_smooth(method=lm, formula=y~x) + theme_bw() +
                          xlab("number of basepaired positions in spacer") + 
                          ylab("rate above no activator (RFU/min)") +
                          annotate(geom="text", x=1, y=-15, 
                                   label=paste("cor =", 
                                               round(spacer_structure_cor$estimate, 
                                                     digits=3))) +
                          annotate(geom="text", x=1, y=-20,
                                   label=paste("p =", 
                                               round(spacer_structure_cor$p.value, 
                                                     digits=3))) +
                          ggtitle("rate ~ spacer structure")) /
                         (ggplot_violin(guide_rate, x_var="lan_cat", y_var="spacer_bp",
                                        x_lab="% unstructured positions in viral target",
                                        y_lab="number of basepaired positions in spacer") +
                            guides(fill=F) + ggtitle("spacer structure ~ target structure"))) | 
  (ggplot(subset(lan_spacer_bp_coef, coef != "(Intercept)"),
          aes(x=Estimate, y=coef)) + geom_point() + theme_bw() +
     geom_errorbar(aes(xmin=Estimate+qnorm(0.05)*Std..Error,
                       xmax=Estimate-qnorm(0.05)*Std..Error)) +
     geom_vline(xintercept=0, linetype="dashed") + 
     xlab("regression coefficient") + ylab("") +
     ggtitle("rate ~ spacer structure + target structure")) | 
  (ggplot_violin(lan_spacer_bp_data, x_var="lan_cat", y_var="rate", 
                 x_lab="% unstructured positions in viral target") +
     guides(fill=F) + ggtitle("rate ~ target structure", 
                              subtitle="controlled for spacer structure") +
     annotate(geom="text", x="50-95%", y=80,
              label=paste("* p =", round(lan_spacer_bp_wilcox$p.value, digits=3))))

# Sun (2021)
guide_rate$sun_invivo_cat <- cut(guide_rate$sun_invivo,
                                 breaks=quantile(guide_rate$sun_invivo, 
                                                 probs=c(0, 0.25, 0.50, 0.75, 1)),
                                 include.lowest=T,
                                 labels=c("Q1", "Q2", "Q3", "Q4"))
guide_rate$sun_invitro_cat <- cut(guide_rate$sun_invitro,
                                 breaks=quantile(guide_rate$sun_invitro, 
                                                 probs=c(0, 0.25, 0.50, 0.75, 1)),
                                 include.lowest=T,
                                 labels=c("Q1", "Q2", "Q3", "Q4"))
sun_plot <- (ggplot_violin(guide_rate, x_var="sun_invivo_cat", 
                           x_lab="quartiles of mean(icSHAPE) across viral target") +
               guides(fill=F) + ggtitle("Sun (2021)", subtitle="in vivo")) +
  (ggplot_violin(guide_rate, x_var="sun_invitro_cat", 
                 x_lab="quartiles of mean(icSHAPE) across viral target") + 
     guides(fill=F) + ggtitle("", subtitle="in vitro"))

# Huston (2021)
guide_rate$huston_cat <- cut(guide_rate$huston,
                             breaks=c(0, 0.2, 0.4, 0.6, 0.8, 1),
                             include.lowest=T,
                             labels=c("0-19%", "20-39%", "40-59%", "60-79%", "80-100%"))
huston_fit <- data.frame(summary(lm(Estimate ~ huston_cat, 
                                    data=guide_rate))$coefficients)
huston_wilcox <- with(guide_rate,
                      wilcox.test(Estimate[huston_cat=="0-19%"], 
                                  Estimate[huston_cat=="40-59%"]))
huston_plot <- (ggplot(guide_rate, aes(x=huston, y=Estimate)) + 
    geom_point() + geom_smooth(method=loess, formula=y~x) + theme_bw() + 
    xlab("% unbound positions in viral target") + 
    ylab("rate above no activator (RFU/min)") +
    ggtitle("Huston (2021)", subtitle="Viral RNA secondary structure")) + 
  (ggplot_violin(guide_rate, x_var="huston_cat", 
                 x_lab="% unbound positions in viral target") + 
     guides(fill=F) + annotate(geom="text", x="40-59%", y=80,
                               label=paste("* p =", round(huston_wilcox$p.value, digits=3))))

lan_plot
lan_spacer_bp_plot
huston_plot
manfredonia_plot + sun_plot + plot_layout(widths=c(1,2))
```

* Correlation between viral structure as annotated by Lan (2020)
    + But no correlation with other *in vivo* annotations (Sun and Huston)

```{r rate_by_viral_structure_intersection, fig.height=4}
(ggplot(guide_rate, aes(x=sun_invivo, y=huston, col=lan, shape=lan_sun_huston)) + 
  geom_point(alpha=0.75) + theme_bw() + 
  xlab("Sun (in vivo)") + ylab("Huston") + labs(col="Lan", shape="Intersection")) +
  (ggplot_violin(guide_rate, x_var="lan_sun_huston") + 
     guides(fill=F))
```

### k-means clustering of in vivo structures (Lan, Sun, Huston)

```{r rate_by_viral_structure_kmeans, fig.height=8}
invivo_kmeans_2_wilcox <- wilcox.test(Estimate ~ invivo_kmeans_2, 
                                      data=guide_rate)
invivo_kmeans_3_wilcox <- wilcox.test(Estimate ~ invivo_kmeans_3,
                                      data=subset(guide_rate,
                                                  invivo_kmeans_3 %in% c("structured", "unstructured")))

# (fviz_cluster(invivo_kmeans_3,
#               data=guide_features[, c("lan", "sun_invivo", "huston")],
#               geom="point", ellipse.type="confidence") +
#     theme_bw() + guides(col=F, shape=F) +
#     scale_fill_discrete(labels=c("structured", "intermediate", "unstructured")[invivo_kmeans_3_labels]))
# (ggplot(guide_rate, aes(x=invivo_kmeans_3, y=Estimate)) + theme_bw() +
#         geom_violin(aes(fill=invivo_kmeans_3), alpha=0.5, 
#                     draw_quantiles=c(0.25, 0.75), linetype="dashed") +
#         geom_violin(fill="transparent", draw_quantiles=0.5) + 
#         geom_dotplot(binaxis="y", stackdir="center", binwidth=1) + guides(fill=F) + 
#         xlab("") + ylab("rate above no activator (RFU/min)") +
#         ggtitle("Guide activity v. viral genome structure", subtitle="k means: 3 clusters") +
#         annotate(geom="text", x= "unstructured", y=80, 
#                  label=paste("p =", round(invivo_kmeans_3_wilcox$p.value, 
#                                           digits=3))))

((fviz_nbclust(guide_features[, c("lan", "sun_invivo", "huston")], 
               kmeans, method="silhouette")) + 
    (fviz_cluster(invivo_kmeans_2, 
                  data=guide_features[, c("lan", "sun_invivo", "huston")], 
                  geom="point", ellipse.type="confidence") + 
       theme_bw() + guides(col=F, shape=F) + 
       scale_fill_discrete(labels=c("structured", "unstructured")[invivo_kmeans_2_labels]))) /
  (ggplot_violin(guide_rate, x_var="invivo_kmeans_2") + guides(fill=F) + 
     ggtitle("k-means clustering of in vivo viral structure (Lan, Sun, Huston)",
             subtitle="k = 2") +
     annotate(geom="text", x="unstructured", y=80,
              label=paste("p =", round(invivo_kmeans_2_wilcox$p.value,
                                       digits=3))))
```

### Restriction to spacer seed region (positions 8-12 on spacer)

```{r rate_by_viral_structure_seed, fig.height=8}
# Manfredonia (2020)
guide_rate$manfredonia_seed_cat <- ifelse(guide_rate$manfredonia_seed==1,
                                          "unstructured",
                                          ifelse(guide_rate$manfredonia_seed==0,
                                                 "structured",
                                                 "intermediate"))
guide_rate$manfredonia_seed_cat <- factor(guide_rate$manfredonia_seed_cat,
                                          levels=c("structured", "intermediate", "unstructured"))
manfredonia_seed_plot <- ggplot_violin(guide_rate, x_var="manfredonia_seed_cat") + 
  ggtitle("Manfredonia (2020): spacer seed region") + guides(fill=F)

# Lan (2020)
guide_rate$lan_seed_cat <- ifelse(guide_rate$lan_seed==1,
                                  "unstructured",
                                  ifelse(guide_rate$lan_seed==0,
                                         "structured",
                                         "intermediate"))
guide_rate$lan_seed_cat <- factor(guide_rate$lan_seed_cat,
                                  levels=c("structured", "intermediate", "unstructured"))
lan_seed_plot <- ggplot_violin(guide_rate, x_var="lan_seed_cat") + 
  ggtitle("Lan (2020): spacer seed region") + guides(fill=F)

# Sun (2021): in vitro
sun_invitro_seed_cat <- kmeans(guide_rate$sun_invitro_seed, 2)
guide_rate$sun_invitro_seed_cat <- factor(sun_invitro_seed_cat$cluster)
levels(guide_rate$sun_invitro_seed_cat) <- c("structured", "unstructured")[rank(sun_invitro_seed_cat$centers)]
sun_invitro_seed_plot <- ggplot_violin(guide_rate, x_var="sun_invitro_seed_cat") + 
  ggtitle("Sun (2020): in vitro: spacer seed region") + guides(fill=F)

# Sun (2021): in vivo
sun_invivo_seed_cat <- kmeans(guide_rate$sun_invivo_seed, 2)
guide_rate$sun_invivo_seed_cat <- factor(sun_invivo_seed_cat$cluster)
levels(guide_rate$sun_invivo_seed_cat) <- c("structured", "unstructured")[rank(sun_invivo_seed_cat$centers)]
sun_invivo_seed_plot <- ggplot_violin(guide_rate, x_var="sun_invivo_seed_cat") +
  ggtitle("Sun (2020): in vivo: spacer seed region") + guides(fill=F)

# Huston (2021)
guide_rate$huston_seed <- as.factor(guide_rate$huston_seed)
huston_seed_plot <- ggplot_violin(guide_rate, x_var="huston_seed") + 
  ggtitle("Huston (2020): spacer seed region") + guides(fill=F)

manfredonia_seed_plot + lan_seed_plot + 
  sun_invitro_seed_plot + sun_invivo_seed_plot +
  huston_seed_plot + plot_spacer() + plot_layout(ncol=2)
```

## Position regression

```{r position_regression, fig.height=10}
structure_features <- c("spacer_only", "manfredonia", "lan", "sun_invivo", "sun_invitro", "huston")
spacer_viral_structure <- lapply(structure_features,
                                 function(feature) {
                                   # wrangle data.frame from regression
                                   spacer_seq <- sapply(guide_rate$start,
                                                        function(x) viral_genome$spacer_base[x:(x+19)])
                                   if(feature == "spacer_only") {
                                     regression_data <- data.frame(guide_rate$Estimate,
                                                                   t(spacer_seq),
                                                                   stringsAsFactors=T)
                                     colnames(regression_data) <- c("rate", paste0("spacer_", 1:20))
                                   } else {
                                     viral_structure <- sapply(guide_rate$start,
                                                               function(x) {
                                                                 viral_structure <- viral_genome[[feature]][x:(x+19)]
                                                                 if(grepl("sun", feature)) {
                                                                   viral_structure <- as.numeric(viral_structure)
                                                                 }
                                                                 if(feature == "huston") {
                                                                   viral_structure <- as.character((viral_structure == 0))
                                                                 }
                                                                 return(viral_structure)
                                                               })
                                     regression_data <- data.frame(guide_rate$Estimate,
                                                                   t(spacer_seq),
                                                                   t(viral_structure),
                                                                   stringsAsFactors=T)
                                     colnames(regression_data) <- c("rate",
                                                                    paste0("spacer_", 1:20),
                                                                    paste0("structure_", 1:20))
                                   }
                                   position_features <- colnames(regression_data)[-1]
                                   regression_data <- subset(regression_data,
                                                             nchar(guide_rate$spacer)==20)
                                   regression_data <- mltools::one_hot(data.table::as.data.table(regression_data))
                                   # compute full model
                                   full_model <- lm(rate ~ ., data=regression_data)
                                   # compute positional models
                                   position_models <- lapply(position_features,
                                                             function(x) {
                                                               if(grepl("structure", x) & grepl("sun", feature)) {
                                                                 tmp_features <- x
                                                               } else {
                                                                 tmp_features <- grep(paste0(x, "_"),
                                                                                    colnames(regression_data),
                                                                                    value=T)
                                                               }
                                                               tmp_data <- subset(regression_data,
                                                                                  select=colnames(regression_data) %in% 
                                                                                    c("rate", tmp_features))
                                                               lm("rate ~ .", data=tmp_data)
                                                             })
                                   position_models_r2 <- data.frame(position_features,
                                                                    sapply(position_models,
                                                                           function(x) summary(x)$r.squared),
                                                                    matrix(unlist(strsplit(position_features,
                                                                                           split="_")),
                                                                           ncol=2, byrow=T))
                                   colnames(position_models_r2) <- c("var", "r2", "type", "pos")
                                   position_models_r2$pos <- factor(position_models_r2$pos, 
                                                                    levels=1:20)
                                   feature_plot <- ggplot(position_models_r2, 
                                          aes(x=pos, y=r2, fill=type)) + 
                                     geom_col(position="stack") + 
                                     theme_bw() + ylim(0, 0.06) +
                                     xlab("position in spacer") + 
                                     ylab(expr(r^2)) + 
                                     ggtitle("Rate ~ spacer sequence + viral genome structure",
                                             subtitle=bquote("full model" ~ r^2 == 
                                                               .(round(summary(full_model)$r.squared, 
                                                                       digits=3)) ~ 
                                                               "| structure:"~.(feature)))
                                   return(list(data=regression_data,
                                               full_model=full_model,
                                               position_models=position_models,
                                               position_models_r2=position_models_r2,
                                               plot=feature_plot))
                                 })
names(spacer_viral_structure) <- structure_features
spacer_viral_structure$spacer_only$plot <- spacer_viral_structure$spacer_only$plot + 
  ggtitle("Rate ~ spacer sequence", 
          subtitle=bquote("full model" ~ r^2 == 
                            .(round(summary(spacer_viral_structure$spacer_only$full_model)$r.squared,
                                   digits=3))))
wrap_plots(lapply(spacer_viral_structure, function(x) x$plot), ncol=1)
spacer_viral_structure$spacer_only$plot / spacer_viral_structure$lan$plot

spacer_only_coef <- data.frame(summary(spacer_viral_structure$spacer_only$full_model)$coefficients)[-1,]
lan_structure_coef <- data.frame(summary(spacer_viral_structure$lan$full_model)$coefficients)[-1,]
lan_structure_coef <- subset(lan_structure_coef, 
                             grepl("structure", rownames(lan_structure_coef)))
lan_coef <- rbind(spacer_only_coef, lan_structure_coef)
lan_coef <- data.frame(lan_coef, 
                       matrix(unlist(strsplit(rownames(lan_coef), split="_")), 
                              ncol=3, byrow=T))
colnames(lan_coef) <- c("estimate", "std_error", "t", "pvalue", 
                        "type", "position", "feature")
lan_coef$position <- factor(lan_coef$position, levels=1:20)
(ggplot(subset(lan_coef, type=="spacer"), 
        aes(x=position, y=estimate, col=feature)) +
    geom_point() + theme_bw() + geom_hline(yintercept=0) + 
    ggtitle("sequence")) / 
  (ggplot(subset(lan_coef, type=="structure"), 
          aes(x=position, y=estimate, col=feature)) + 
     geom_point() + theme_bw() + geom_hline(yintercept=0) +
     ggtitle("viral structure (Lan 2020)"))
```

* "Unstructured" label from Lan (2020) is most predictive
* Also informative:
    + Base-paired-ness from Huston (2021) in position 1
    + *in vitro* icSHAPE score from Sun (2021) in positions 6-9
* Spacer sequence is most informative in positions 2, 13, 17, 18

### Likelihood ratio test (spacer-only vs. spacer+structure): 

```{r lrtest_spaceronly_spacerlan}
lmtest::lrtest(spacer_viral_structure$spacer_only$full_model,
                                          spacer_viral_structure$lan$full_model)
```

## Elastic net regression on all features

```{r position_regression_elasticnet}
regression_data <- lapply(colnames(viral_genome)[-1],
                          function(feature) {
                            t(sapply(guide_rate$start,
                                     function(x) viral_genome[[feature]][x:(x+19)]))
                          })
regression_data <- data.frame(do.call(cbind, regression_data))
colnames(regression_data) <- paste(rep(colnames(viral_genome)[-1], each=20),
                                   rep(1:20, times=length(colnames(viral_genome)[-1])),
                                   sep="_")
for(x in grep("base", colnames(regression_data), value=T)) {
  regression_data[[x]] <- factor(regression_data[[x]], 
                                 levels=c("A", "T", "C", "G"))
}
for(x in grep("manfredonia", colnames(regression_data), value=T)) {
  regression_data[[x]] <- factor(regression_data[[x]], 
                                 levels=c(".", "lowShannon/highSHAPE"))
}
for(x in grep("lan", colnames(regression_data), value=T)) {
  regression_data[[x]][regression_data[[x]]=="both"] <- "."
  regression_data[[x]] <- factor(regression_data[[x]],
                                 levels=c(".", "unstructured", "structured"))
}
for(x in grep("sun", colnames(regression_data), value=T)) {
  regression_data[[x]] <- as.numeric(regression_data[[x]])
}
for(x in grep("huston", colnames(regression_data), value=T)) {
  regression_data[[x]] <- factor(regression_data[[x]] == 0, 
                                 levels=c("TRUE", "FALSE"))
}

# 1. convert predictors into one-hot encoded matrix
regression_data.table <- mltools::one_hot(data.table::as.data.table(regression_data))
# 2. pre-compute folds for cross-validation
folds <- cut(seq(nrow(regression_data)), breaks=10, labels=F)
folds <- sample(folds)
# 3. run cv.glmnet over pre-specified alpha
alphas <- seq(0, 1, by=0.1)
cv_glmnet_fits <- lapply(alphas,
                         function(alpha) {
                           glmnet::cv.glmnet(as.matrix(regression_data.table),
                                             guide_rate$Estimate,
                                             foldid=folds, alpha=alpha)
                         })
cv_glmnet_mse <- lapply(seq_along(cv_glmnet_fits),
                        function(x) {
                          tmp_fit <- cv_glmnet_fits[[x]]
                          data.frame(lambda=tmp_fit$lambda,
                                     mse=tmp_fit$cvm,
                                     mse_sd=tmp_fit$cvsd,
                                     mse_upper=tmp_fit$cvup,
                                     mse_lower=tmp_fit$cvlo,
                                     coef_zero=tmp_fit$nzero,
                                     alpha=alphas[x])
                        })
cv_glmnet_mse <- do.call(rbind, cv_glmnet_mse)
alpha_plot <- ggplot(cv_glmnet_mse, aes(x=lambda, y=mse, col=as.factor(alpha))) + 
  geom_point(size=0.5) + geom_line() + scale_x_log10() + scale_y_log10() + 
  xlab(expr(lambda)) + ylab("MSE") + labs(col=expr(alpha)) + theme_bw()
best_alpha <- cv_glmnet_mse$alpha[which.min(cv_glmnet_mse$mse)]
best_lambda <- cv_glmnet_mse$lambda[which.min(cv_glmnet_mse$mse)]
best_fit <- cv_glmnet_fits[[which(alphas==best_alpha)]]$glmnet.fit
best_fit_coefs <- coef(best_fit, s=best_lambda)[,1]
```

* $\alpha=0$: ridge regression; $\alpha=1$: LASSO
* best fit: 
    + $\alpha =$ `r best_alpha` 
    + $\lambda =$ `r best_lambda`
    + regression coefficients: `r best_fit_coefs[best_fit_coefs!=0]`

## Sequence logo

```{r sequence_logo}
allSpacers_freq <- compute_nt_freq(subset(guide_rate$spacer,
                                          nchar(guide_rate$spacer)==20))

topSpacers_freq <- compute_nt_freq(subset(guide_rate$spacer,
                                          nchar(guide_rate$spacer) == 20 &
                                            rank(-guide_rate$Estimate) <= 20))
topSpacers_info <- calculate_relative_info(topSpacers_freq, allSpacers_freq)

bottomSpacers_freq <- compute_nt_freq(subset(guide_rate$spacer,
                                             nchar(guide_rate$spacer) == 20 &
                                               rank(guide_rate$Estimate) <= 20))
bottomSpacers_info <- calculate_relative_info(bottomSpacers_freq, allSpacers_freq)

(ggseqlogo::ggseqlogo(topSpacers_info, method="custom") + 
  ggtitle("Sequence logo of top 20 spacers")) / 
  (ggseqlogo::ggseqlogo(bottomSpacers_info, method="custom") +
     ggtitle("Sequence logo for bottom 20 spacers"))
```

Top guides:

* G in position 19
* U's in positions 8-9, 11-14

Bottom guides:

* A in position 8
* G's in positions 8-12, 18-20

## Comparison to screening performed in gBlocks

```{r load_gblock_data}
# load round 1 data
gblock_round1_data <- xlsx::read.xlsx(file.path(project_dir,
                                                "NCR_Guide_Data.xlsx"),
                                      sheetName="Data",
                                      rowIndex=1:112)
gblock_round1_data <- subset(gblock_round1_data,
                             NCR.ID %in% guide_rate$NCR.id)
gblock_round1_data$X100.fM.Activator.above.noActivator <- with(gblock_round1_data,
                                                               as.numeric(X100.fM.Activator.Rate) -
                                                                 as.numeric(No.Activator.Rate))

# load round 2, plate 1-2 platemap
gblock_round2_platemap_1_2 <- xlsx::read.xlsx(file.path(project_dir, 
                                                        "Plate1_2_gBlock_384w_replicate_platemap.xlsx"),
                                              sheetIndex=1)
gblock_round2_platemap_1_2 <- subset(gblock_round2_platemap_1_2,
                                     subset=value_type %in% LETTERS[1:16])[, -1]
gblock_round2_platemap_1_2_activator <- data.frame(matrix("_noActivator", 
                                                nrow=nrow(gblock_round2_platemap_1_2),
                                                ncol=ncol(gblock_round2_platemap_1_2)))
gblock_round2_platemap_1_2_activator[, 13:24] <- "_100fM"
gblock_round2_platemap_1_2_activator[c(2, 4, 6), 9] <- "_100fM"
gblock_round2_platemap_1_2_activator[c(2, 4, 6), 11] <- "_100fM"
gblock_round2_platemap_1_2 <- data.frame(sample = paste0(unlist(gblock_round2_platemap_1_2),
                                                       unlist(gblock_round2_platemap_1_2_activator)),
                                         well_384 = paste0(rep(LETTERS[1:16], times=24),
                                                           rep(1:24, each=16)))
gblock_round2_platemap_1_2$sample <- sub(" ", "_", gblock_round2_platemap_1_2$sample)

# load round 2, plate 2-2 platemap
gblock_round2_platemap_2_2 <- xlsx::read.xlsx(file.path(project_dir, 
                                                        "Plate2_2_gBlock_384w_replicate_platemap.xlsx"),
                                              sheetIndex=1)
gblock_round2_platemap_2_2 <- subset(gblock_round2_platemap_2_2,
                                     subset=value_type %in% LETTERS[1:16])[, -1]
gblock_round2_platemap_2_2_activator <- data.frame(matrix("_noActivator", 
                                                nrow=nrow(gblock_round2_platemap_2_2),
                                                ncol=ncol(gblock_round2_platemap_2_2)))
gblock_round2_platemap_2_2_activator[, 13:24] <- "_100fM"
gblock_round2_platemap_2_2_activator[c(2, 4, 6), 9] <- "_100fM"
gblock_round2_platemap_2_2_activator[c(2, 4, 6), 11] <- "_100fM"
gblock_round2_platemap_2_2 <- data.frame(sample = paste0(unlist(gblock_round2_platemap_2_2),
                                                       unlist(gblock_round2_platemap_2_2_activator)),
                                         well_384 = paste0(rep(LETTERS[1:16], times=24),
                                                           rep(1:24, each=16)))
gblock_round2_platemap_2_2$sample <- sub(" ", "_", gblock_round2_platemap_2_2$sample)

# load round 2 data
gblock_data_dir <- file.path(project_dir, "100fM_gblock_data")
gblock_data_fnames <- list.files(gblock_data_dir)
gblock_data <- lapply(gblock_data_fnames,
                      function(x) {
                        tmp_plate <- sub("_100fM_Plate", "",
                                         sub("-", "_",
                                             sub(".xlsx", "", x)))
                        row_indices <- 55:115
                        tmp_data <- xlsx::read.xlsx(file.path(gblock_data_dir, x),
                                                    sheetIndex=1, rowIndex=row_indices, header=T)
                        tmp_data <- lapply(seq(nrow(tmp_data)),
                                           function(x) {
                                             data.frame(tmp_data[x, 1:3],
                                                        unlist(tmp_data[x, 4:387]),
                                                        colnames(tmp_data)[4:387],
                                                        row.names=NULL)
                                           })
                        tmp_data <- do.call(rbind, tmp_data)
                        colnames(tmp_data) <- c("cycle_number", "time", "temp", "signal", "well_384")
                        if(tmp_plate %in% paste0("gBlock", c("1_1", "2_1"))) {
                          well_96 <- mapping_384_to_96$well_96[match(tmp_data$well_384,
                                                                     mapping_384_to_96$well_384)]
                          tmp_platemap <- plate_maps[[sub("gBlock", "GS", tmp_plate)]]
                          tmp_data$sample <- tmp_platemap$sample[match(well_96,
                                                                       tmp_platemap$well_96)]
                        } else {
                          if(tmp_plate=="gBlock1_2") {
                            tmp_platemap <- gblock_round2_platemap_1_2
                          } else {
                            tmp_platemap <- gblock_round2_platemap_2_2
                          }
                          tmp_data$sample <- tmp_platemap$sample[match(tmp_data$well_384, tmp_platemap$well_384)]
                        }
                        tmp_data$plate <- tmp_plate
                        return(tmp_data)
                      })
gblock_data <- do.call(rbind, gblock_data)
gblock_data$sample[grepl("NA", gblock_data$sample)] <- "empty"
gblock_data$guide_id <- sapply(gblock_data$sample,
                               function(x) {
                                 ifelse(grepl("_", x),
                                        sub(" ", "_", sub("_.*", "", x)),
                                        "empty")
                               })
gblock_data$guide_id[gblock_data$guide_id=="612"] <- "612_Control"
gblock_data$guide_id[gblock_data$guide_id=="No"] <- "No_Protein"
gblock_data$activator <- factor(sapply(gblock_data$sample,
                                       function(x) {
                                         ifelse(grepl("_", x),
                                                sub(".*_", "", x),
                                                "empty")
                                       }), 
                                levels=c("100fM", "noActivator", "empty"))
gblock_data$cycle_number <- as.numeric(gblock_data$cycle_number)
gblock_data$signal <- as.numeric(gblock_data$signal)
gblock_data$time <- gblock_data$time/60 # convert from seconds to minutes
gblock_bkgd_empty <- mean(subset(gblock_data, sample=="empty")$signal, na.rm=T)
gblock_data$signal_bkgdSubtract <- gblock_data$signal - gblock_bkgd_empty
gblock_data <- subset(gblock_data, !(gblock_data$sample == "empty"))
gblock_data$activator <- droplevels(gblock_data$activator)
gblock_data <- subset(gblock_data, !is.na(signal))

# compute gblock rates
gblock_guide_ids <- unique(gblock_data$guide_id)
gblock_guide_ids <- gblock_guide_ids[!(gblock_guide_ids %in% c("612_Control", "No_Protein"))]
gblock_results <- lapply(gblock_guide_ids,
                         function(x) {
                           tmp <- try(analyze_guide(gblock_data, guide=x, time_start=15,
                                                    signal="signal_bkgdSubtract",
                                                    mixed_model=mixed),
                                      silent=T)
                           if(class(tmp) == "try-error") {
                             print(paste0("mixed model failed: NCR_", x))
                             tmp <- analyze_guide(gblock_data, guide=x, time_start=15, 
                                                  signal="signal_bkgdSubtract",
                                                  mixed_model=F)
                             return(tmp)
                           } else {
                             return(tmp)
                           }
                         })
names(gblock_results) <- gblock_guide_ids
gblock_plots <- lapply(seq_along(gblock_results),
                       function(x) {
                         guide_id <- paste0("NCR_", names(gblock_results)[x])
                         which_row <- match(guide_id, guide_features$NCR.id)
                         has_hairpin <- guide_features$has_crRNA_hairpin[which_row]
                         spacer_bp <- guide_features$crRNA_spacer_basepairs[which_row]
                         plot_subtitle <- paste(ifelse(has_hairpin, "has hairpin", "no hairpin"), "|",
                                                spacer_bp, "basepaired positions in spacer")
                         gblock_results[[x]]$plot + ggtitle(guide_id, subtitle=plot_subtitle)
                       })
names(gblock_plots) <- gblock_guide_ids
gblock_coef <- lapply(gblock_results, function(x) x$coef)
gblock_coef <- do.call(rbind, gblock_coef)
gblock_rate <- subset(gblock_coef, grepl(":", gblock_coef$coef))

# annotate guide_rate with gblock rate
guide_rate$gblock_rate_round1 <- NA
guide_rate$gblock_rate_round1[match(gblock_round1_data$NCR.ID,
                                    guide_rate$NCR.id)] <- gblock_round1_data$X100.fM.Activator.above.noActivator
guide_rate$gblock_rate_round2[match(paste0("NCR_", gblock_rate$guide_id),
                                    guide_rate$NCR.id)] <- gblock_rate$Estimate
(ggplot(subset(guide_rate, !is.na(gblock_rate_round1)), 
        aes(x=Estimate, y=gblock_rate_round1)) + 
    geom_point(alpha=0.5) + theme_bw() + ggtitle("gBlock: round 1") +
    geom_smooth(method=lm, formula=y~x) +
    xlab("viral RNA: rate above no activator (RFU/min)") + 
    ylab("gBlocks: rate above no activator (RFU/min)") +
    annotate(geom="text", x=-Inf, y=Inf, 
             label=paste("cor =", round(with(guide_rate, 
                                             cor(Estimate, gblock_rate_round1, 
                                                 use="pairwise.complete.obs",
                                                 method="spearman")),
                                        digits=3)),
             hjust=-0.25, vjust=2, col="red")) + 
  (ggplot(guide_rate, aes(x=Estimate, y=gblock_rate_round2)) +
     geom_point(alpha=0.5) + theme_bw() + 
     geom_smooth(method=lm, formula=y~x) +
     xlab("viral RNA: rate above no activator (RFU/min)") +
     ylab("gBlocks: rate above no activator (RFU/min)") +
     ggtitle("gBlock: round 2") + ylim(0, 170) + 
     annotate(geom="text", x=-Inf, y=Inf, 
              label=paste("cor =", round(with(guide_rate, 
                                              cor(Estimate, gblock_rate_round2, 
                                                  use="pairwise.complete.obs",
                                                  method="spearman")),
                                         digits=3)),
              hjust=-0.25, vjust=2, col="red")) + 
  (ggplot(subset(guide_rate, !is.na(gblock_rate_round1)),
          aes(x=gblock_rate_round1, y=gblock_rate_round2)) + 
     geom_point(alpha=0.5) + geom_smooth(method="lm", formula=y~x) + 
     theme_bw() + xlab("round 1 (RFU/min)") + ylab("round 2 (RFU/min)") + 
     ggtitle("gBlocks: round 1 vs. round 2") + 
     annotate(geom="text", x=-Inf, y=Inf,
              label=paste("cor =", round(with(guide_rate,
                                              cor(gblock_rate_round1, 
                                                  gblock_rate_round2,
                                                  use="pairwise.complete.obs",
                                                  method="spearman")),
                                         digits=3)),
              hjust=-0.25, vjust=2, col="red"))
```

gBlock round 2 outlier:

```{r gblock_NCR_1352}
gblock_plots[["1352"]]
```

# Supplementary: all traces

## "Good" guides

```{r suppl_good_guide, }
plots_by_group$`good guide`
```

## "Bad" guides: maintains hairpin, 4 basepaired positions in spacer

```{r suppl_hasHairpin_4bp}
plots_by_group$`hasHairpin/4bp`
```

## "Bad" guides: no hairpin, 4 basepaired positions in spacer

```{r suppl_noHairpin_4bp}
plots_by_group$`noHairpin/4bp`
```

## "Bad" guides: maintains hairpin, 10 basepaired positions in spacer

```{r suppl_hasHairpin_10bp}
plots_by_group$`hasHairpin/10bp`
```

## "Bad" guides: no hairpin, 10 basepaired positions in spacer

```{r suppl_noHairpin_10bp}
plots_by_group$`noHairpin/10bp`
```

## "Bad" guides: maintains hairpin, 16 basepaired positions in spacer

```{r suppl_hasHairpin_16bp}
plots_by_group$`hasHairpin/16bp`
```

## "Bad" guides: no hairpin, 16 basepaired positions in spacer

```{r suppl_noHairpin_16bp}
plots_by_group$`noHairpin/16bp`
```

## NCR_1344

```{r suppl_NCR_1344}
all_plots[["1344"]]
```

## gBlock traces

```{r suppl_gBlocks}
gblock_plots
```